In [115]:
import pandas as pd

In [116]:
df=pd.read_csv("amazon_prime_titles.csv")

In [117]:
df.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')

In [118]:
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,The Grand Seduction,Don McKellar,"Brendan Gleeson, Taylor Kitsch, Gordon Pinsent",Canada,"March 30, 2021",2014,NaN,113 min,"Comedy, Drama",A small fishing village must procure a local d...
1,s2,Movie,Take Care Good Night,Girish Joshi,"Mahesh Manjrekar, Abhay Mahajan, Sachin Khedekar",India,"March 30, 2021",2018,13+,110 min,"Drama, International",A Metro Family decides to fight a Cyber Crimin...
2,s3,Movie,Secrets of Deception,Josh Webber,"Tom Sizemore, Lorenzo Lamas, Robert LaSardo, R...",United States,"March 30, 2021",2017,NaN,74 min,"Action, Drama, Suspense",After a man discovers his wife is cheating on ...
3,s4,Movie,Pink: Staying True,Sonia Anderson,"Interviews with: Pink, Adele, Beyoncé, Britney...",United States,"March 30, 2021",2014,NaN,69 min,Documentary,"Pink breaks the mold once again, bringing her ..."
4,s5,Movie,Monster Maker,Giles Foster,"Harry Dean Stanton, Kieran O'Brien, George Cos...",United Kingdom,"March 30, 2021",1989,NaN,45 min,"Drama, Fantasy",Teenage Matt Banting wants to work with a famo...


In [119]:
df.nunique()

show_id         9668
type               2
title           9668
director        5773
cast            7927
country           86
date_added        84
release_year     100
rating            24
duration         219
listed_in        518
description     9414
dtype: int64

In [120]:
df.shape

(9668, 12)

In [121]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9668 entries, 0 to 9667
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       9668 non-null   object
 1   type          9668 non-null   object
 2   title         9668 non-null   object
 3   director      7585 non-null   object
 4   cast          8435 non-null   object
 5   country       672 non-null    object
 6   date_added    155 non-null    object
 7   release_year  9668 non-null   int64 
 8   rating        9331 non-null   object
 9   duration      9668 non-null   object
 10  listed_in     9668 non-null   object
 11  description   9668 non-null   object
dtypes: int64(1), object(11)
memory usage: 906.5+ KB


In [122]:
df.describe()

,release_year
count,9668.000000
mean,2008.341849
std,18.922482
min,1920.000000
25%,2007.000000
50%,2016.000000
75%,2019.000000
max,2021.000000


In [123]:
df["director"].nunique()

5773

In [124]:
# Check how many rows have missing ratings
print(df['rating'].isnull().sum())


337


In [125]:
# Drop rows where rating is missing
df = df.dropna(subset=['rating'])

# Verify
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 9331 entries, 1 to 9667
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       9331 non-null   object
 1   type          9331 non-null   object
 2   title         9331 non-null   object
 3   director      7255 non-null   object
 4   cast          8123 non-null   object
 5   country       657 non-null    object
 6   date_added    140 non-null    object
 7   release_year  9331 non-null   int64 
 8   rating        9331 non-null   object
 9   duration      9331 non-null   object
 10  listed_in     9331 non-null   object
 11  description   9331 non-null   object
dtypes: int64(1), object(11)
memory usage: 947.7+ KB
None


In [126]:
#Tried Creating Dictionary But Witnessed An Error
unique_values_dict = {
    col: df[col].unique().tolist()
    for col in df.columns
}

print(unique_values_dict)

IOPub data rate exceeded.
The Jupyter server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--ServerApp.iopub_data_rate_limit`.

Current values:
ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
ServerApp.rate_limit_window=3.0 (secs)



In [127]:
#Creating Unique Values Json File For Analysis
import json

with open("unique_values.json", "w", encoding="utf-8") as f:
    json.dump(unique_values_dict, f, indent=4, default=str)

In [128]:
#Replacing Few Ratings Which Aren't Useful 
rating_map = {
    '16': '16+',
    'AGES_16_': '16+',
    '18+': '18+',
    'AGES_18_': '18+',
    'ALL': 'ALL',
    'ALL_AGES': 'ALL',
    'NR': 'NOT_RATED',
    'UNRATED': 'NOT_RATED',
    'TV-NR': 'NOT_RATED',
    'NOT_RATE': 'NOT_RATED'
}

df['rating'] = df['rating'].replace(rating_map)

In [129]:
#Originally 25 rating reduced to 17
print(sorted(df['rating'].dropna().unique()))

['13+', '16+', '18+', '7+', 'ALL', 'G', 'NC-17', 'NOT_RATED', 'PG', 'PG-13', 'R', 'TV-14', 'TV-G', 'TV-MA', 'TV-PG', 'TV-Y', 'TV-Y7']


In [130]:
df["rating"].isna().sum()

np.int64(0)

In [131]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9331 entries, 1 to 9667
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       9331 non-null   object
 1   type          9331 non-null   object
 2   title         9331 non-null   object
 3   director      7255 non-null   object
 4   cast          8123 non-null   object
 5   country       657 non-null    object
 6   date_added    140 non-null    object
 7   release_year  9331 non-null   int64 
 8   rating        9331 non-null   object
 9   duration      9331 non-null   object
 10  listed_in     9331 non-null   object
 11  description   9331 non-null   object
dtypes: int64(1), object(11)
memory usage: 947.7+ KB


In [132]:
#Now I Feel We Have Obtained A Dataset Where We Have Appropriate Ratings.

In [133]:
'''
Checking The Genre Of Each Movie/Show ➡️ Observation : Observed A Lot Many Genres Then Actually Present As It Contains A 
                                                        Mixture Of Different Genres.Now Let's Brainstorm What We Can Do 🤔.
'''
print(sorted(df['listed_in'].unique()))

['Action', 'Action, Adventure', 'Action, Adventure, Animation', 'Action, Adventure, Anime', 'Action, Adventure, Comedy', 'Action, Adventure, Documentary', 'Action, Adventure, Drama', 'Action, Adventure, Faith and Spirituality', 'Action, Adventure, Fantasy', 'Action, Adventure, Horror', 'Action, Adventure, International', 'Action, Adventure, Kids', 'Action, Adventure, Science Fiction', 'Action, Adventure, Special Interest', 'Action, Adventure, Sports', 'Action, Adventure, Suspense', 'Action, Adventure, Western', 'Action, Animation', 'Action, Animation, Anime', 'Action, Animation, Drama', 'Action, Animation, Fantasy', 'Action, Animation, Kids', 'Action, Animation, Science Fiction', 'Action, Anime', 'Action, Anime, Comedy', 'Action, Anime, Drama', 'Action, Anime, Fantasy', 'Action, Anime, International', 'Action, Anime, Kids', 'Action, Arthouse, Arts, Entertainment, and Culture', 'Action, Arthouse, Drama', 'Action, Arthouse, Horror', 'Action, Arts, Entertainment, and Culture', 'Action, Ar

In [134]:
#Exporting The Actual Unique Values Of All Genres From 'listed_in' Column
all_genres = set()

for genres in df['listed_in'].dropna():
    for genre in genres.split(','):
        all_genres.add(genre.strip())

all_genres = sorted(all_genres)

print(all_genres)
print(f"\nTotal unique genres: {len(all_genres)}")

['Action', 'Adventure', 'Animation', 'Anime', 'Arthouse', 'Arts', 'Comedy', 'Documentary', 'Drama', 'Entertainment', 'Faith and Spirituality', 'Fantasy', 'Fitness', 'Historical', 'Horror', 'International', 'Kids', 'LGBTQ', 'Military and War', 'Music Videos and Concerts', 'Romance', 'Science Fiction', 'Special Interest', 'Sports', 'Suspense', 'TV Shows', 'Talk Show and Variety', 'Unscripted', 'Western', 'Young Adult Audience', 'and Culture']

Total unique genres: 31


In [135]:
#Renaming listed_in Column To Genre
df.rename(columns={'listed_in': 'Genre'}, inplace=True)

In [136]:
print(df.columns)

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'Genre', 'description'],
      dtype='object')


In [137]:
genre_dummies = (
    df['Genre']
    .str.get_dummies(sep=',')
)

genre_dummies.columns = (
    genre_dummies.columns
    .str.strip()
    .str.replace(' ', '_')
)

genre_dummies = genre_dummies.add_prefix('Genre_')

In [91]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9331 entries, 1 to 9667
Data columns (total 69 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   show_id                          9331 non-null   object
 1   type                             9331 non-null   object
 2   title                            9331 non-null   object
 3   director                         7255 non-null   object
 4   cast                             8123 non-null   object
 5   country                          657 non-null    object
 6   date_added                       140 non-null    object
 7   release_year                     9331 non-null   int64 
 8   rating                           9331 non-null   object
 9   duration                         9331 non-null   object
 10  Genre                            9331 non-null   object
 11  description                      9331 non-null   object
 12  Genre_Adventure                  9331 n

In [92]:
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,...,Genre_Romance,Genre_Science Fiction,Genre_Special Interest,Genre_Sports,Genre_Suspense,Genre_TV Shows,Genre_Talk Show and Variety,Genre_Unscripted,Genre_Western,Genre_Young Adult Audience
1,s2,Movie,Take Care Good Night,Girish Joshi,"Mahesh Manjrekar, Abhay Mahajan, Sachin Khedekar",India,"March 30, 2021",2018,13+,110 min,...,0,0,0,0,0,0,0,0,0,0
16,s17,Movie,Zoombies,Glenn Miller,"Marcus Anderson, Kaiwi Lyman, Andrew Asper",NaN,NaN,2016,13+,87 min,...,0,0,0,0,0,0,0,0,0,0
17,s18,TV Show,Zoo Babies,NaN,Narrator - Gillian Barlett,NaN,NaN,2008,ALL,1 Season,...,0,0,0,0,0,0,0,0,0,0
18,s19,TV Show,Zoë Coombs Marr: Bossy Bottom,NaN,Zoë Coombs Marr,NaN,NaN,2020,18+,1 Season,...,0,0,0,0,0,0,0,0,0,0
19,s20,Movie,Zoe,Drake Doremus,"Ewan McGregor, Léa Seydoux, Theo James, Mirand...",NaN,NaN,2018,R,104 min,...,0,1,0,0,0,0,0,0,0,0


In [95]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9331 entries, 1 to 9667
Data columns (total 69 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   show_id                          9331 non-null   object
 1   type                             9331 non-null   object
 2   title                            9331 non-null   object
 3   director                         7255 non-null   object
 4   cast                             8123 non-null   object
 5   country                          657 non-null    object
 6   date_added                       140 non-null    object
 7   release_year                     9331 non-null   int64 
 8   rating                           9331 non-null   object
 9   duration                         9331 non-null   object
 10  Genre                            9331 non-null   object
 11  description                      9331 non-null   object
 12  Genre_Adventure                  9331 n

In [138]:
df = df.loc[:, ~df.columns.duplicated()]

In [140]:
genre_dummies = df['Genre'].str.get_dummies(sep=',')

# clean column names
genre_dummies.columns = genre_dummies.columns.str.strip()

# add prefix
genre_dummies = genre_dummies.add_prefix('Genre_')

# merge
df = pd.concat([df, genre_dummies], axis=1)

In [142]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9331 entries, 1 to 9667
Data columns (total 69 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   show_id                          9331 non-null   object
 1   type                             9331 non-null   object
 2   title                            9331 non-null   object
 3   director                         7255 non-null   object
 4   cast                             8123 non-null   object
 5   country                          657 non-null    object
 6   date_added                       140 non-null    object
 7   release_year                     9331 non-null   int64 
 8   rating                           9331 non-null   object
 9   duration                         9331 non-null   object
 10  Genre                            9331 non-null   object
 11  description                      9331 non-null   object
 12  Genre_Adventure                  9331 n

In [143]:
df['Genre'] = df['Genre'].fillna('')

df['Genre'] = df['Genre'].apply(
    lambda x: ','.join([i.strip() for i in x.split(',')])
)

In [144]:
df = df.drop(columns=[col for col in df.columns if col.startswith("Genre_")], errors='ignore')

In [145]:
genre_dummies = df['Genre'].str.get_dummies(sep=',')

# clean column names
genre_dummies.columns = genre_dummies.columns.str.strip()

# add prefix
genre_dummies = genre_dummies.add_prefix('Genre_')

# merge
df = pd.concat([df, genre_dummies], axis=1)

In [146]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9331 entries, 1 to 9667
Data columns (total 43 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   show_id                          9331 non-null   object
 1   type                             9331 non-null   object
 2   title                            9331 non-null   object
 3   director                         7255 non-null   object
 4   cast                             8123 non-null   object
 5   country                          657 non-null    object
 6   date_added                       140 non-null    object
 7   release_year                     9331 non-null   int64 
 8   rating                           9331 non-null   object
 9   duration                         9331 non-null   object
 10  Genre                            9331 non-null   object
 11  description                      9331 non-null   object
 12  Genre_Action                     9331 n

In [147]:
print([col for col in df.columns if col.startswith("Genre_")])
print("Total genre columns:", len([col for col in df.columns if col.startswith("Genre_")]))

['Genre_Action', 'Genre_Adventure', 'Genre_Animation', 'Genre_Anime', 'Genre_Arthouse', 'Genre_Arts', 'Genre_Comedy', 'Genre_Documentary', 'Genre_Drama', 'Genre_Entertainment', 'Genre_Faith and Spirituality', 'Genre_Fantasy', 'Genre_Fitness', 'Genre_Historical', 'Genre_Horror', 'Genre_International', 'Genre_Kids', 'Genre_LGBTQ', 'Genre_Military and War', 'Genre_Music Videos and Concerts', 'Genre_Romance', 'Genre_Science Fiction', 'Genre_Special Interest', 'Genre_Sports', 'Genre_Suspense', 'Genre_TV Shows', 'Genre_Talk Show and Variety', 'Genre_Unscripted', 'Genre_Western', 'Genre_Young Adult Audience', 'Genre_and Culture']
Total genre columns: 31


In [149]:
#Now Clearing Leading And Trailing Space From The Values Of All The Columns 
df = df.apply(lambda col: col.str.strip() if col.dtype == "object" else col)

In [150]:
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,...,Genre_Science Fiction,Genre_Special Interest,Genre_Sports,Genre_Suspense,Genre_TV Shows,Genre_Talk Show and Variety,Genre_Unscripted,Genre_Western,Genre_Young Adult Audience,Genre_and Culture
1,s2,Movie,Take Care Good Night,Girish Joshi,"Mahesh Manjrekar, Abhay Mahajan, Sachin Khedekar",India,"March 30, 2021",2018,13+,110 min,...,0,0,0,0,0,0,0,0,0,0
16,s17,Movie,Zoombies,Glenn Miller,"Marcus Anderson, Kaiwi Lyman, Andrew Asper",NaN,NaN,2016,13+,87 min,...,1,0,0,0,0,0,0,0,0,0
17,s18,TV Show,Zoo Babies,NaN,Narrator - Gillian Barlett,NaN,NaN,2008,ALL,1 Season,...,0,1,0,0,0,0,0,0,0,0
18,s19,TV Show,Zoë Coombs Marr: Bossy Bottom,NaN,Zoë Coombs Marr,NaN,NaN,2020,18+,1 Season,...,0,0,0,0,0,1,0,0,0,0
19,s20,Movie,Zoe,Drake Doremus,"Ewan McGregor, Léa Seydoux, Theo James, Mirand...",NaN,NaN,2018,R,104 min,...,1,0,0,0,0,0,0,0,0,0


In [151]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9331 entries, 1 to 9667
Data columns (total 43 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   show_id                          9331 non-null   object
 1   type                             9331 non-null   object
 2   title                            9331 non-null   object
 3   director                         7255 non-null   object
 4   cast                             8123 non-null   object
 5   country                          657 non-null    object
 6   date_added                       140 non-null    object
 7   release_year                     9331 non-null   int64 
 8   rating                           9331 non-null   object
 9   duration                         9331 non-null   object
 10  Genre                            9331 non-null   object
 11  description                      9331 non-null   object
 12  Genre_Action                     9331 n

In [152]:
#Considered removing the date_Added,country as both has a lot many null values
#but keeping it and will not use it for any analysis.so they are useless column for now
df['country'] = df['country'].str.strip()

In [153]:
df['country'].unique()

array(['India', nan, 'Italy', 'United States', 'United Kingdom, France',
       'United States, Italy', 'United States, India',
       'United Kingdom, United States',
       'United States, United Kingdom, Germany', 'Japan',
       'China, United States, United Kingdom',
       'Denmark, United Kingdom, Czech Republic, Netherlands',
       'United Kingdom', 'Canada', 'United States, Ireland',
       'United States, United Kingdom, Canada',
       'United Kingdom, United States, India', 'United Kingdom, India',
       'United States, China', 'Canada, United States',
       'United States, Thailand', 'Spain', 'Brazil, United States, India',
       'Switzerland, India', 'United Kingdom, Canada, United States',
       'Germany', 'Australia', 'Belgium',
       'Chile, Argentina, United States', 'Mexico',
       'Denmark, Sweden, United States', 'Netherlands',
       'United States, United Kingdom', 'New Zealand',
       'United States, Portugal', 'United States, France',
       'Canada, Un

In [154]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9331 entries, 1 to 9667
Data columns (total 43 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   show_id                          9331 non-null   object
 1   type                             9331 non-null   object
 2   title                            9331 non-null   object
 3   director                         7255 non-null   object
 4   cast                             8123 non-null   object
 5   country                          657 non-null    object
 6   date_added                       140 non-null    object
 7   release_year                     9331 non-null   int64 
 8   rating                           9331 non-null   object
 9   duration                         9331 non-null   object
 10  Genre                            9331 non-null   object
 11  description                      9331 non-null   object
 12  Genre_Action                     9331 n

In [155]:
#Removing Rows Containing Null Values Of Cast Column 
df = df.dropna(subset=['cast'])

In [156]:
print(df['cast'].isna().sum())
print(df.shape)

0
(8123, 43)


In [157]:
df.info()
#now one more problem director column has null values aprrox 1500+ which is a bad news but let's keep those

<class 'pandas.core.frame.DataFrame'>
Index: 8123 entries, 1 to 9667
Data columns (total 43 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   show_id                          8123 non-null   object
 1   type                             8123 non-null   object
 2   title                            8123 non-null   object
 3   director                         6645 non-null   object
 4   cast                             8123 non-null   object
 5   country                          640 non-null    object
 6   date_added                       93 non-null     object
 7   release_year                     8123 non-null   int64 
 8   rating                           8123 non-null   object
 9   duration                         8123 non-null   object
 10  Genre                            8123 non-null   object
 11  description                      8123 non-null   object
 12  Genre_Action                     8123 n

In [159]:
df.dropna(subset=['director']).info()
#on droping rows containing null values of director all the columns has almost all values but date_Added has reduced to just 
#1 non null value.but we will still drop the null values of director rows.

<class 'pandas.core.frame.DataFrame'>
Index: 6645 entries, 1 to 9667
Data columns (total 43 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   show_id                          6645 non-null   object
 1   type                             6645 non-null   object
 2   title                            6645 non-null   object
 3   director                         6645 non-null   object
 4   cast                             6645 non-null   object
 5   country                          538 non-null    object
 6   date_added                       1 non-null      object
 7   release_year                     6645 non-null   int64 
 8   rating                           6645 non-null   object
 9   duration                         6645 non-null   object
 10  Genre                            6645 non-null   object
 11  description                      6645 non-null   object
 12  Genre_Action                     6645 n

In [160]:
df = df.dropna(subset=['director'])

In [161]:
print(df['director'].isna().sum())
print(df.shape)

0
(6645, 43)


In [162]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6645 entries, 1 to 9667
Data columns (total 43 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   show_id                          6645 non-null   object
 1   type                             6645 non-null   object
 2   title                            6645 non-null   object
 3   director                         6645 non-null   object
 4   cast                             6645 non-null   object
 5   country                          538 non-null    object
 6   date_added                       1 non-null      object
 7   release_year                     6645 non-null   int64 
 8   rating                           6645 non-null   object
 9   duration                         6645 non-null   object
 10  Genre                            6645 non-null   object
 11  description                      6645 non-null   object
 12  Genre_Action                     6645 n

In [165]:
#Renaming Genre_and Culture to Genre_Culture
df.rename(columns={'Genre_and Culture': 'Genre_Culture'}, inplace=True)

In [166]:
df.to_csv("cleaned_dataset.csv", index=False)

🌞What We Did Today :-

---

# ➡️ Dataset Cleaning & Transformation Summary

## 1. Initial Data Understanding

* Explored dataset using `df.info()`
* Identified key columns:

  * `rating`, `listed_in`, `cast`, `director`, `country`, `date_added`
* Noticed missing values and multi-value columns (like genres)

---

## 2. Handling Missing Values

* **Dropped rows with missing `rating`**

  ```python
  df = df.dropna(subset=['rating'])
  ```
* **Decision on `director` and `cast`:**

  * Preferred keeping data instead of losing rows
  * Optionally replaced missing values with `"Unknown"`

---

## 3. Cleaning Text Data

* Removed leading/trailing spaces in:

  * Column names
  * All string/object values

```python
df.columns = df.columns.str.strip()

df = df.apply(lambda col: col.str.strip() if col.dtype == "object" else col)
```

---

## 4. Genre Column Transformation (Major Step)

### Problem:

* `listed_in` had combined values like:

  ```
  Action, Adventure, Comedy
  ```
* This created incorrect “unique values” (512 combinations)

### Solution:

* Split genres into individual values
* Cleaned spaces and inconsistencies

```python
df['Genre'] = df['Genre'].apply(lambda x: ','.join([i.strip() for i in x.split(',')]))
```

---

## 5. One-Hot Encoding for Genre

* Converted multi-label genres into binary columns

```python
genre_dummies = df['Genre'].str.get_dummies(sep=',')
genre_dummies.columns = genre_dummies.columns.str.strip()
genre_dummies = genre_dummies.add_prefix('Genre_')
df = pd.concat([df, genre_dummies], axis=1)
```

### Fixes applied:

* Removed duplicate genre columns
* Fixed spacing issues causing duplicate columns

---

## 6. Column Standardization

* Renamed:

  * `listed_in` → `Genre`
* Fixed inconsistent column name:

  * `Genre_and Culture` → `Genre_Culture`

```python
df.rename(columns={'Genre_and Culture': 'Genre_Culture'}, inplace=True)
```

---

## 7. Handling Sparse Columns

### `date_added`

* Only ~140 values
* Kept but marked as optional

### `country`

* ~657 values
* Decided to keep for potential analysis

---

## 8. Exporting Final Dataset

* Saved cleaned dataset for Power BI / analysis

```python
df.to_csv("cleaned_dataset.csv", index=False)
```

---

# 🎯 Final Outcome

I now have:

* Cleaned dataset (6645 rows)
* Standardized text data
* One-hot encoded genres
* Reduced noise and inconsistencies
* Power BI–ready structure

---

# 🚀 What I Built (Important)

You essentially transformed a raw dataset into:

> A structured analytics-ready dataset with feature engineering (Genre encoding + cleaning + normalization)

---


In [ ]:
"""
Conclusion:
1.After Cleaning The Data We Have Obtained A Dataset Containing 6645 rows which is 68.7% of total rows (Original Dataset : 9668 Rows)
2.Learned About How Complex It Becomes Sometimes Even Just Forgetting To Remove Leading/Trail Spaces.I Encounterred The Problem While Creating Dummies
        Of Genre.
3.Now,Have To Analyse The Data using PowerBI.
"""